solid_attenuator/ioc_lfe_at2l0_calc/calculator.py

```diff
@@ -167,6 +187,73 @@ def get_best_config(all_transmissions: typing.List[float],
     return floor_config if mode == ConfigMode.Floor else ceil_config
 
 
+def get_best_config_with_material_priority(
+        materials: typing.List[str],
+        transmissions: typing.List[float],
+        material_order: typing.List[str],
+        t_des: float,
+        ) -> Config:
+    """
+    Inserting filters based on the material order provided, get the best
+    possible filter configuration.
+
```

Comment:
This should have a short blurb explaining that the materials with lower priority will not be used at all until the materials with higher priority are all inserted

Reply:
Noted this in 5186917

Reply:
I said this because it was in the attached issues, but I don't see where in the code this is accomplished... How does this work? All the tests seem to pass...

Reply:
Very good question - I didn't explain this nearly as well as I should have. Let me try here before reformatting and attempting to put into code. How's this sound?

The original algorithm
--------------------
The generic "best configuration" algorithm does its best to choose the right filters to target the requested transmission.
It does this by enumerating all possible filter configurations, doing a matrix multiplication to determine the transmission values for all of those configurations. Figuring out the best configuration is then a matter of matching the closest transmission value (i.e., `argmin(abs(all_transmissions - target_transmission))`.  This makes no assumptions as to how the filters are laid out and is as generic as possible.

Now, if all filters of every material type are used with the above algorithm, some silicon filters may be inserted prior to diamond filters. This is not good, as the silicon filters may be damaged. We need a new algorithm to get around that.

This PR
--------
The proposed algorithm breaks up the problem into 2 calls to the original algorithm, once per material.

Using the example of AT2L0, with 8 diamond filters and 10 silicon filters:

(1) Apply the original algorithm _only_ for the diamond filters. Due to how the filters are physically configured, the diamond filters can get very close to the required transmission up to a certain value (`t_all_diamond_inserted` to `1.0`). This works because the filter thicknesses were carefully specified: each filter is 2x the previous filter in thickness. 

As an aside, this means you could treat each filter as binary value, spanning the range  `t_all_diamond_inserted <= t_des <= 1.0` in 2^num_filters = 2^8 steps (i.e., `(1 - t_all_diamond_inserted) / 2 ** 8`, which is `< 0.4%` per step)

Let's call the result from step (1) a transmission value of` transmission_1`.

(2) Now,
(a) For desired transmission within `t_all_diamond_inserted <= t_des <= 1.0`, the algorithm will have pretty much gotten the requested transmission as close as needed.
(b) For desired transmission `0 <= t_des < t_all_diamond_inserted`, the algorithm will have already selected all diamond filters. There's more left to do in the second round as we can still get closer to `t_des`, when calculating which silicon filters to insert.

(3) As normalized transmission values are multiplicative (50% of 50% transmission would be 25%), our second round tries to find the right configuration of silicon filters for `t_des / transmission_1`. 
For values in the range noted by (2a), `t_des / transmission1` will be close to 1 - so no filters will be inserted.
For values in the range noted by (2b), `t_des / transmission1` will likely be something actionable.

(4) Assembling the two configurations from (1) and (3) gives a final configuration, specifying all filter states.

Reply:
Ok, so it stops at step `2a` and avoids putting in `Si` in these cases by forcing the calculation to slightly overshoot the attenuation (undershoot the transmission) such that adding additional `Si` filters will only make us get further away from the goal attenuation? And that is why I was confused by the code which doesn't have any additional stopping parameter here? I think this was my misunderstanding point.

Reply:
Yes, the choice of 'ceiling' in step 2a overshoots the transmission, ensuring that no silicon filters are inserted unless necessary. That's a good thing to clarify in the docs as well - thanks for looking through this.

Reply:
_Actually_, the actually should have been `floor` to safely insert the silicon filters, if for some reason diamond filters stop working (i.e., fail in the out position).  Brute forcing energy ranges and transmission values confirms the above.

This does change the test suite cases and makes the results not match up 1-to-1 with what Matt had determined previously. I'll have to think about how to rework them in a sensible way; perhaps by adding the brute-force verification of configurations based on energy, desired transmission, and non-working filters to the test suite.

Reply:
That's an annoying edge case for sure

Reply:
OK, all of the above is captured in the docs now, slightly modified from the above due to some algorithm changes I made today.